### DISCUSSION TITLE
How to attach file for each specific email ?

### DISCUSSION THREAD

**Discussion by huynd26** [anymail/django-anymail#303] [Q&A]
Reporting an error? It's helpful to know:
I have some changes in ics file for each recipient and want to make sure that each recipient gets their own ics file instead of one attachment for all emails like below.

```python
    msg = AnymailMessage(
        subject=subject, body=body, from_email=from_email, to=list_emails
    )
    msg.attach("meeting_invite.ics", meeting_invite, "text/calendar")
````

- Anymail version : 7.0.0
- ESP (Mailgun, SendGrid, etc.) : SparkPost = 1.3.6
- Your ANYMAIL settings (change secrets to "redacted") : 
```ANYMAIL = {
    "SPARKPOST_API_KEY": os.environ.get("SPARKPOST_API_KEY"),
    "SEND_DEFAULTS": {
        "esp_extra": {
            "transactional": True,
            "return_path": "mail@bounces.project.com",
            "ip_pool": "project",
        },
        "track_clicks": True,
    },
    "WEBHOOK_SECRET": os.environ.get("ANYMAIL_WEBHOOK_SECRET"),
}
EMAIL_BACKEND = "anymail.backends.sparkpost.EmailBackend"```
- Versions of Django, requests, python
Django=3.2.18, python=3.9.15+

**Comment by medmunds** [comment]
The easiest solution is probably to loop over your list and send a separate message to each recipient, with whatever changes you need for their ics attachment:

```python
for to_email in list_emails:
    msg = AnymailMessage(
        subject=subject, body=body, from_email=from_email, to=[to_email]
    )
    msg.attach(
        "meeting_invite.ics", 
        # Apply whatever changes you want for this recipient:
        customize_ics_for_invitee(meeting_invite, to_email), 
        "text/calendar")
    msg.send()
```

(Anymail has some batch sending features that can use a single Django `EmailMessage` as a "recipe" for creating different custom messages for each recipient, but the things you can change in that recipe are very limited. Attachments are *not* included in Anymail's current per-recipient batch send options, due to limitations in most ESP's batch sending APIs.)

(As a general rule, if you want to send different emails to different recipients, it's usually simpler to just do that with a loop in your code like shown above, rather than trying to use Anymail's batch sending.)

**Reply by huynd26** [reply]
Thank you, I'm using the way with a loop over my list, but is this a good solution? If I have to send the to thousands of recipients, whether it is ok for performance?

**Reply by medmunds** [reply]
I wouldn't expect any performance problems with looping over your recipient list vs. making a single batch send call to SparkPost's API.

https connection overhead shouldn't be a concern: the requests library Anymail uses to call SparkPost's API should cache and reuse the connection. But if it does seem to be a problem, you could use Django's `connection.send_messages()` function to send a list of `AnymailMessage` objects using a single https connection: https://docs.djangoproject.com/en/4.1/topics/email/#topics-sending-multiple-emails

If you're regularly sending thousands of messages at a time, the bigger concern is error handling: you're going to run into network glitches (or SparkPost's rate limits) at some point. The loop approach will have simpler error detection and recovery than any of the batch sending approaches, because you're dealing with exactly one recipient at a time.
